### ISSUE TITLE
make instruction / script to setup dev environment on NixOS

### ISSUE THREAD

**Issue by caryoscelus** [zeronet-conservancy/zeronet-conservancy#189]
```
ERROR:root:Unhandled exception: libstdc++.so.6: cannot open shared object file: No such file or directory
Traceback (most recent call last):
  File "./zeronet.py", line 52, in main
    import main
  File "src/main.py", line 13, in <module>
    import gevent
  File "venv/lib/python3.9/site-packages/gevent/__init__.py", line 86, in <module>
    from gevent._hub_local import get_hub
  File "venv/lib/python3.9/site-packages/gevent/_hub_local.py", line 101, in <module>
    import_c_accel(globals(), 'gevent.__hub_local')
  File "venv/lib/python3.9/site-packages/gevent/_util.py", line 148, in import_c_accel
    mod = importlib.import_module(cname)
  File "/nix/store/j652sgyb3137c19v5vy5ziaarw5k5bf6-python3-3.9.15/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "src/gevent/_hub_local.py", line 1, in init gevent._gevent_c_hub_local
  File "venv/lib/python3.9/site-packages/greenlet/__init__.py", line 29, in <module>
    from ._greenlet import _C_API # pylint:disable=no-name-in-module
ImportError: libstdc++.so.6: cannot open shared object file: No such file or directory
```

@fgaz can you help ?

**Comment by fgaz** [comment]
TL;DR: don't use venv, use `nix-shell '<nixpkgs>' -A zeronet-conservancy` or write your own `shell.nix`/`default.nix`.

NixOS doesn't use [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard), so any  code compiled in "normal" distros (such as what is contained in that pypi package) won't work in NixOS and vice versa. FHS can be emulated through a [buildFHSUserEnv](https://nixos.org/manual/nixpkgs/stable/index.html#sec-fhs-environments) wrapper ([steam-run](https://nixos.org/manual/nixpkgs/stable/index.html#sec-steam-run) is a pre-made one that works for most stuff), but ideally one would do it the nix way by writing a nix expression that provides a suitable environment without emulating other distros.

In general this can be done with [mkShell](https://nixos.org/manual/nixpkgs/stable/index.html#sec-pkgs-mkShell). If you paste the example from the page i linked in a `shell.nix` file then run `nix-shell` you'll be dropped in a shell with the specified dependencies.

In the nixpkgs manual there are also instructions for [every lanuage ecosystem known to nixpkgs](https://nixos.org/manual/nixpkgs/stable/index.html#chap-language-support). [Python](https://nixos.org/manual/nixpkgs/stable/index.html#python) has a rather large section.

...fortunately, since zeronet-conservancy is already present in nixpkgs, you can just tell nix to give you a shell for that: `nix-shell '<nixpkgs>' -A zeronet-conservancy` (or if you use the experimental flakes interface `nix develop nixpkgs#zeronet-conservancy`)

**Comment by caryoscelus** [comment]
@fgaz thanks for quick answer ! when i have time , i'll try to learn NixOS way of development . what's interesting is why it used to work and no longer works , but i guess that's not the point

> ..fortunately, since zeronet-conservancy is already present in nixpkgs

yeah ^_^ but i wanted to launch another instance of dev version..

anyway , i'll reuse this issue to track README/scripts to help setting up dev environment on NixOS

**Comment by fgaz** [comment]
> it used to work and no longer works

Then it could also be that a previous pip invocation built a native library and linked it against some nix package that got garbage collected since then. A different problem but still fairly unique to nix.

> yeah ^_^ but i wanted to launch another instance of dev version..

But you can! One of the main uses of `nix-shell` is development. `nix-shell '<nixpkgs>' -A zeronet-conservancy` opens a shell that you can use to develop (and run) `zeronet-conservancy`. Whether it's the version in nixpkgs or the development version it doesn't matter (as long as you don't alter the dependencies. then you'd need to modify the nix file or create a new one). So you can just `cd` to your clone of the repo, open the `nix-shell`, and run the development version with `python zeronet.py`

**Comment by caryoscelus** [comment]
> But you can! One of the main uses of nix-shell is development. nix-shell '<nixpkgs>' -A zeronet-conservancy opens a shell that you can use to develop (and run) zeronet-conservancy. Whether it's the version in nixpkgs or the development version it doesn't matter (as long as you don't alter the dependencies. then you'd need to modify the nix file or create a new one). So you can just cd to your clone of the repo, open the nix-shell, and run the development version with python zeronet.py

ah , i see , cool . will try it out and add to README then . thanks for the explanation !
